upup/pkg/fi/cloudup/dotasks/loadbalancer.go

```diff
@@ -0,0 +1,188 @@
+	cloud := c.Cloud.(*digitalocean.Cloud)
+	lbService := cloud.LoadBalancers()
+
+	if fi.StringValue(lb.ID) != "" {
+		loadbalancer, _, err := lbService.Get(context.TODO(), fi.StringValue(lb.ID))
+
+		if err != nil {
+			return nil, fmt.Errorf("load balancer service get request returned error %v", err)
+		}
+
+		// do another check to double-sure we are talking to the right load balancer.
+		return &LoadBalancer{
+			Name:      fi.String(loadbalancer.Name),
+			ID:        fi.String(loadbalancer.ID),
+			Lifecycle: lb.Lifecycle,
+			Region:    fi.String(loadbalancer.Region.Slug),
+		}, nil
+	}
+
+	// Loadbalancer = nil if not found
+	return nil, nil
+}
+
+func (lb *LoadBalancer) Run(c *fi.Context) error {
+	return fi.DefaultDeltaRunMethod(lb, c)
+}
+
+func (_ *LoadBalancer) CheckChanges(a, e, changes *LoadBalancer) error {
+	if a != nil {
+		if changes.Name != nil {
+			return fi.CannotChangeField("Name")
+		}
+		if changes.ID != nil {
+			return fi.CannotChangeField("ID")
+		}
+		if changes.Region != nil {
+			return fi.CannotChangeField("Region")
+		}
+	} else {
+		if e.Name == nil {
+			return fi.RequiredField("Name")
+		}
+		if e.Region == nil {
+			return fi.RequiredField("Region")
+		}
+	}
+	return nil
+}
+
+func (_ *LoadBalancer) RenderDO(t *do.DOAPITarget, a, e, changes *LoadBalancer) error {
+
+	Rules := []godo.ForwardingRule{
+		{
+			EntryProtocol:  "https",
+			EntryPort:      443,
+			TargetProtocol: "https",
+			TargetPort:     443,
+			TlsPassthrough: true,
+		},
+		{
+			EntryProtocol:  "http",
+			EntryPort:      80,
+			TargetProtocol: "http",
+			TargetPort:     80,
+		},
+	}
+
+	HealthCheck := &godo.HealthCheck{
+		Protocol:               "tcp",
+		Port:                   443,
+		Path:                   "",
+		CheckIntervalSeconds:   60,
+		ResponseTimeoutSeconds: 5,
+		UnhealthyThreshold:     3,
+		HealthyThreshold:       5,
+	}
+
+	klog.V(10).Infof("Creating load balancer for DO")
+
+	loadBalancerService := t.Cloud.LoadBalancers()
+	loadbalancer, _, err := loadBalancerService.Create(context.TODO(), &godo.LoadBalancerRequest{
+		Name:            fi.StringValue(e.Name),
+		Region:          fi.StringValue(e.Region),
+		Tag:             fi.StringValue(e.DropletTag),
+		ForwardingRules: Rules,
+		HealthCheck:     HealthCheck,
+	})
+
+	if err != nil {
+		klog.Errorf("Error creating load balancer with Name=%s, Error=%v", fi.StringValue(e.Name), err)
+		return err
+	}
+
+	e.ID = fi.String(loadbalancer.ID)
+	e.IPAddress = fi.String(loadbalancer.IP) // This will be empty on create, but will be filled later on FindIPAddress invokation.
+
+	return nil
+}
+
+func (lb *LoadBalancer) FindIPAddress(c *fi.Context) (*string, error) {
+	cloud := c.Cloud.(*digitalocean.Cloud)
+	loadBalancerService := cloud.LoadBalancers()
+
+	klog.V(10).Infof("Find IP address for load balancer ID=%s", fi.StringValue(lb.ID))
+	loadBalancer, _, err := loadBalancerService.Get(context.TODO(), fi.StringValue(lb.ID))
```

Comment:
Are we certain at this point that we have an LB ID? Genuine question as I don't know the kops work flow in this context that well.

Reply:
Yes @timoreimann - we are certain we have an LB ID, but not certain about the LB IP Address. FindAddress is only called after RenderDO is invoked - we create a new LB Request in RenderDO that returns a LB ID. We use this LB ID to find if the IP Address is available.